common/rpc/rpc.go

```diff
@@ -235,3 +253,103 @@ func (d *RPCFactory) dial(hostName string, tlsClientConfig *tls.Config) *grpc.Cl
 func (d *RPCFactory) GetTLSConfigProvider() encryption.TLSConfigProvider {
 	return d.tlsFactory
 }
+
+// CreateLocalFrontendHTTPClient gets or creates a cached frontend client.
+func (d *RPCFactory) CreateLocalFrontendHTTPClient() (*common.FrontendHTTPClient, error) {
+	return d.localFrontendClient()
+}
+
+// createLocalFrontendHTTPClient creates an HTTP client for communicating with the frontend.
+// It uses either the provided frontendURL or membership to resolve the frontend address.
+func (d *RPCFactory) createLocalFrontendHTTPClient() (*common.FrontendHTTPClient, error) {
+	// dialer and transport field values copied from http.DefaultTransport.
+	dialer := &net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}
+	transport := &http.Transport{
+		Proxy:                 http.ProxyFromEnvironment,
+		DialContext:           dialer.DialContext,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+	client := http.Client{}
+
+	// Default to http unless TLS is configured.
+	scheme := "http"
+	if d.frontendTLSConfig != nil {
+		transport.TLSClientConfig = d.frontendTLSConfig
+		scheme = "https"
+	}
+
+	var address string
+	service, err := extractServiceFromURL(d.frontendHTTPURL)
+	if err != nil {
+		return nil, err
+	}
+	if service != "" {
+		// The URL instructs us to resolve via membership for frontend or internal-frontend.
+		r, err := d.monitor.GetResolver(primitives.ServiceName(service))
+		if err != nil {
+			return nil, err
+		}
+		client.Transport = &roundTripper{
+			resolver:   r,
+			underlying: transport,
+			config:     d.config,
+		}
+		address = "internal" // This will be replaced by the roundTripper
+	} else {
+		// Use the URL as-is and leave the transport unmodified.
+		client.Transport = transport
+		address = d.frontendHTTPURL
+	}
+
+	return &common.FrontendHTTPClient{
+		Client:  client,
+		Address: address,
+		Scheme:  scheme,
+	}, nil
+}
+
+type roundTripper struct {
+	resolver   membership.ServiceResolver
+	underlying http.RoundTripper
+	config     *config.RPC
+}
+
+func (rt *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Pick a frontend host at random.
+	members := rt.resolver.AvailableMembers()
+	if len(members) == 0 {
+		return nil, serviceerror.NewUnavailable("no frontend host to route request to")
+	}
+	idx := rand.Intn(len(members))
+	member := members[idx]
+
+	// Replace port with the HTTP port.
+	host, _, err := net.SplitHostPort(member.Identity())
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract port from frontend member: %w", err)
+	}
+	address := fmt.Sprintf("%s:%d", host, rt.config.HTTPPort)
+
+	// Replace request's host.
+	req.URL.Host = address
```

Comment:
FWIW the comments on `http.RoundTripper` say not to modify the request. I think it should not be an issue for our use case since we don't expect the same request to be reused, but it is technically not recommended since a RoundTripper should be safe for use by multiple goroutines.

Reply:
I know, I remember our discussion. I think it's fine though given the testing I've done.